File: go/Zadanie3/hashTable.go
package main

import (
	"fmt"
	"strconv"
)

type ChainingHashTable struct {
	lists []*ForwardListTwo // Храним списки напрямую в срезе
	size  int
}

func NewChainingHashTable(tableSize int) *ChainingHashTable {
	lists := make([]*ForwardListTwo, tableSize)

	for i := 0; i < tableSize; i++ {
		lists[i] = newForwardListTwo() // Создаем список для каждой ячейки
	}

	return &ChainingHashTable{
		lists: lists,
		size:  tableSize,
	}
}

func (ht *ChainingHashTable) hash(key int) int {
	if key < 0 {
		key = -key
	}
	return key % ht.size
}

func (ht *ChainingHashTable) getList(index int) *ForwardListTwo {
	if index < 0 || index >= ht.size {
		return nil
	}
	return ht.lists[index] // Возвращаем существующий список
}

func (ht *ChainingHashTable) clearList(list *ForwardListTwo) {
	if list == nil {
		return
	}

	for list.head != nil {
		deleteNodeHeadTwo(list)
	}
}

func (ht *ChainingHashTable) Insert(key int) {
	index := ht.hash(key)
	list := ht.getList(index)
	if list != nil {
		addNodeTailTwo(list, strconv.Itoa(key))
	}
}

func (ht *ChainingHashTable) Search(key int) bool {
	index := ht.hash(key)
	list := ht.getList(index)
	if list == nil {
		return false
	}
	return findNodeIndexTwo(list, strconv.Itoa(key))
}

func (ht *ChainingHashTable) Remove(key int) {
	index := ht.hash(key)
	list := ht.getList(index)
	if list != nil {
		deleteNodeIndexTwo(list, strconv.Itoa(key))
	}
}

func (ht *ChainingHashTable) GetStatistics() (int, int, float64) {
	minLength := 1 << 30
	maxLength := 0
	totalElements := 0
	nonEmptyBuckets := 0

	for i := 0; i < ht.size; i++ {
		list := ht.getList(i)
		if list != nil {
			listSize := countNodesTwo(list)
			if listSize > 0 {
				if listSize < minLength {
					minLength = listSize
				}
				if listSize > maxLength {
					maxLength = listSize
				}
				totalElements += listSize
				nonEmptyBuckets++
			}
		}
	}

	if minLength == 1<<30 {
		minLength = 0
	}

	avgLength := 0.0
	if nonEmptyBuckets > 0 {
		avgLength = float64(totalElements) / float64(nonEmptyBuckets)
	}

	return minLength, maxLength, avgLength
}

func (ht *ChainingHashTable) PrintDetailedStatistics() {
	totalElements := 0
	emptyBuckets := 0
	maxChain := 0
	minChain := 1 << 30

	fmt.Println("Распределение цепочек:")
	for i := 0; i < ht.size; i++ {
		list := ht.getList(i)
		chainLength := 0
		if list != nil {
			chainLength = countNodesTwo(list)
		}

		if chainLength == 0 {
			emptyBuckets++
		}
		if chainLength > maxChain {
			maxChain = chainLength
		}
		if chainLength < minChain && chainLength > 0 {
			minChain = chainLength
		}

		totalElements += chainLength

		if i < 10 {
			fmt.Printf("Ячейка %d: %d элементов\n", i, chainLength)
		}
	}

	if minChain == 1<<30 {
		minChain = 0
	}

	fmt.Println("\nОбщая статистика:")
	fmt.Printf("Минимальная длина непустой цепочки: %d\n", minChain)
	fmt.Printf("Максимальная длина цепочки: %d\n", maxChain)
	fmt.Printf("Пустых ячеек: %d из %d\n", emptyBuckets, ht.size)
	fmt.Printf("Общее количество элементов: %d\n", totalElements)
	fmt.Printf("Коэффициент заполнения: %.2f\n", float64(totalElements)/float64(ht.size))
}

// Добавляем метод GetList для использования в Set
func (ht *ChainingHashTable) GetList(index int) *ForwardListTwo {
	return ht.getList(index)
}

File: с++/Zadanie2/Zadanie2/hash_table.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <climits>
#include "hash_table.h"

using namespace std;

// Хеш-функция: вычисляет индекс по ключу
int ChainingHashTable::hash(int key) const {
    return abs(key) % size;  // Используем модуль для равномерного распределения
}

// Получение цепочки по индексу
ForwardListTwo* ChainingHashTable::getList(int index) const {
    if (index < 0 || index >= size) return nullptr;  // Проверка границ

    string ptrStr = get(*table, index);
    if (ptrStr.empty()) return nullptr;

    // Преобразование строки обратно в указатель
    return reinterpret_cast<ForwardListTwo*>(stoll(ptrStr));
}

// Очистка списка (освобождение памяти всех узлов)
void ChainingHashTable::clearList(ForwardListTwo* list) {
    if (!list) return;

    // Последовательное удаление всех узлов
    while (list->head != nullptr) {
        deleteNodeHeadTwo(list);
    }
    delete list;  // Освобождение самой структуры списка
}

// Конструктор хеш-таблицы
ChainingHashTable::ChainingHashTable(int tableSize) : size(tableSize) {
    table = new Array;
    init(*table, size);  // Инициализация массива

    // Создание пустых списков для каждой ячейки
    for (int i = 0; i < size; i++) {
        ForwardListTwo* newList = new ForwardListTwo;
        newList->head = nullptr;
        newList->tail = nullptr;
        add(*table, "");

        // Сохранение указателя на список как строку в массиве
        set(*table, i, to_string(reinterpret_cast<long long>(newList)));
    }
}

// Деструктор хеш-таблицы
ChainingHashTable::~ChainingHashTable() {
    // Очистка всех цепочек
    for (int i = 0; i < size; i++) {
        ForwardListTwo* list = getList(i);
        clearList(list);
    }

    // Освобождение памяти массива
    destroy(*table);
    delete table;
}

// Вставка элемента в хеш-таблицу
void ChainingHashTable::insert(int key) {
    int index = hash(key);  // Вычисление индекса
    ForwardListTwo* list = getList(index);
    if (list) {
        addNodeTailTwo(list, to_string(key));  // Добавление в конец цепочки
    }
}

// Поиск элемента в хеш-таблице
bool ChainingHashTable::search(int key) const {
    int index = hash(key);
    ForwardListTwo* list = getList(index);
    if (!list) return false;

    // Поиск в цепочке
    return findNodeIndexTwo(list, to_string(key));
}

// Получение статистики по хеш-таблице
void ChainingHashTable::getStatistics(int& minLength, int& maxLength, double& avgLength) const {
    minLength = INT_MAX;
    maxLength = 0;
    int totalElements = 0;
    int nonEmptyBuckets = 0;

    // Анализ всех цепочек
    for (int i = 0; i < size; i++) {
        ForwardListTwo* list = getList(i);
        if (list) {
            int listSize = countNodesTwo(*list);
            if (listSize > 0) {
                // Обновление минимальной и максимальной длины
                if (listSize < minLength) minLength = listSize;
                if (listSize > maxLength) maxLength = listSize;
                totalElements += listSize;
                nonEmptyBuckets++;
            }
        }
    }

    // Корректировка значений для пустой таблицы
    if (minLength == INT_MAX) minLength = 0;

    // Вычисление средней длины непустых цепочек
    avgLength = nonEmptyBuckets > 0 ? static_cast<double>(totalElements) / nonEmptyBuckets : 0;
}

// Удаление элемента из хеш-таблицы
void ChainingHashTable::remove(int key) {
    int index = hash(key);
    ForwardListTwo* list = getList(index);
    if (list) {
        deleteNodeIndexTwo(list, to_string(key));  // Удаление из цепочки
    }
}

// Вывод детальной статистики хеш-таблицы
void ChainingHashTable::printDetailedStatistics() const {
    int totalElements = 0;
    int emptyBuckets = 0;
    int maxChain = 0;
    int minChain = INT_MAX;

    cout << "Распределение цепочек:\n";

    // Анализ всех цепочек
    for (int i = 0; i < size; i++) {
        ForwardListTwo* list = getList(i);
        int chainLength = list ? countNodesTwo(*list) : 0;

        // Сбор статистики
        if (chainLength == 0) emptyBuckets++;
        if (chainLength > maxChain) maxChain = chainLength;
        if (chainLength < minChain && chainLength > 0) minChain = chainLength;

        totalElements += chainLength;

        // Вывод первых 10 цепочек для примера
        if (i < 10) {
            cout << "Ячейка " << i << ": " << chainLength << " элементов\n";
        }
    }

    // Вывод общей статистики
    cout << "\nОбщая статистика:\n";
    cout << "Минимальная длина непустой цепочки: " << (minChain == INT_MAX ? 0 : minChain) << "\n";
    cout << "Максимальная длина цепочки: " << maxChain << "\n";
    cout << "Пустых ячеек: " << emptyBuckets << " из " << size << "\n";
    cout << "Общее количество элементов: " << totalElements << "\n";
    cout << "Коэффициент заполнения: " << static_cast<double>(totalElements) / size << "\n";
}